### ISSUE TITLE
S3 URL Endpoint

### ISSUE THREAD

**Issue by sparrowjack63** [IQSS/dataverse#4690]
Hi all,

Dataverse seems to use awscli to put data on S3 storage,  but unfortunately, there is no way to put a specific S3 endpoint URL in the config file of awscli.
The only and normal way to use an other S3 endpoint URL with awscl is to put it as a parameter on every awscli command like this  : awscl s3 --endpoint-url=https://mys3compatiblestorage.org ...

Could you please add the opportunity to add the parameter "--endpoint-url=URL" to every awscli command used by dataverse ?

It will be very usefull for every users who have S3 compatible storage (like ceph or scality) and want to use it with dataverse.

Thanks!

**Comment by pdurbin** [comment]
@sparrowjack63 thanks for this suggestion. @mbamouni was asking related questions at https://groups.google.com/d/msg/dataverse-community/py0UMJV9lDg/A7Lk0YtsCQAJ and this gives me a better idea of a possible fix. Out of curiosity, would you be interested in making a pull request? 😄

**Comment by sparrowjack63** [comment]
Thanks, 
Unfortunately, I'm not a developper, it would take me too much time to learn how to make this change :(

**Comment by pdurbin** [comment]
@sparrowjack63 ah, good to know that you and @mbamouni work together.

No worries about the pull request. It looks like the only time "awscli" appears in the git repo is in the docs (rather than the code) at https://github.com/IQSS/dataverse/blob/v4.8.6/doc/sphinx-guides/source/installation/config.rst#set-up-access-configuration-via-command-line-tools . But above that section it is described how to manually set up: https://github.com/IQSS/dataverse/blob/v4.8.6/doc/sphinx-guides/source/installation/config.rst#set-up-access-configuration-via-command-line-tools 

I'm confused, though. When setting up manually it seems like one would need the ability to add "endpoint-url" to the "~/.aws/config" file. @sparrowjack63 can you please read through https://github.com/aws/aws-cli/issues/1270 and see if there's anything in there that explains what to do?

**Comment by sparrowjack63** [comment]
I have read https://github.com/aws/aws-cli/issues/1270 but unfortunately, there still no possibility to add s3 endpoint in awscli config file, it's still a pending feature request (for a long time...)

**Comment by pdurbin** [comment]
@sparrowjack63 ok, do you want to try the Swift API for CEPH? Here are the docs:

- http://docs.ceph.com/docs/master/radosgw/swift/
- http://guides.dataverse.org/en/4.8.6/installation/config.html#swift-storage

**Comment by pdurbin** [comment]
@sparrowjack63 sure, I believe all the S3 code is here: https://github.com/IQSS/dataverse/blob/v4.8.6/src/main/java/edu/harvard/iq/dataverse/dataaccess/S3AccessIO.java

If that doesn't help, please let me know what else would be helpful. Please keep asking questions. Thanks!

**Comment by pdurbin** [comment]
@aries-strato thanks for commenting. I'm new to AWS but since you mentioned RDS I thought I'd say that it's already in use by Harvard Dataverse. #4709 is about an upgrade to Postgres 9.6 in general but for anyone on RDS, they'll need to upgrade as well. Come to think of it, Harvard Dataverse is using ELB as well. And the instances run on EC2. Can you perhaps give an example of how the config would ideally look from your perspective?

**Comment by aries-strato** [comment]
Simply adding the field in the .config file would be enough, along with specifying a --no-verify url option as well

[default]
region = cloud1
ec2_endpoint = https://cloud.net/api/ec2
elb_endpoint = https://cloud.net/api/elb
rds_endpoint = https://cloud.net/api/rds
verify_ssl = false

**Comment by pdurbin** [comment]
@aries-strato thanks. Any news from your end? I'm wondering what the status of this issue is. Are you @sparrowjack63 interested in making a pull request?

**Comment by poikilotherm** [comment]
Hey @pdurbin,

this really is a show stopper for our internal test deployment... Our centrally offered Swift Object Store is not GA yet and building a SAIO instance just for testing seems like overkill.

Minio is a very lightweight alternative and should be sufficient for testing purposes (maybe even beyond that?) or some Ceph RADOS S3 gateway...

I just found https://github.com/minio/cookbook/blob/master/docs/aws-sdk-for-java-with-minio.md - it doesn't seem to be that hard to get the custom URL support into the Dataverse S3 storage driver class. 

Wondering if you would like to get a PR for this? I would suggest using a separate storage driver class to make it more distinguishable and less error prone for existing installations... (even so it introduces some code duplication.)

(After having our testbed up and running, we can get to Payara testing... :wink: )

Cheers,
Oliver

**Comment by pdurbin** [comment]
@poikilotherm yes! We're interested in a pull request. I just invited you to join https://github.com/orgs/IQSS/teams/dataverse-readonly so I can assign this issue to you and dragged it to the new "Community Dev" column on our kanban board: https://waffle.io/IQSS/dataverse

We're happy to answer questions here, on https://groups.google.com/forum/#!forum/dataverse-dev or http://chat.dataverse.org (#dataverse on freenode) or during community calls ( https://dataverse.org/community-calls ). Please just get in touch if there's any walk through you need of the code or docs. Thanks!

I'd feel remiss if I didn't say that we used NFS in production for about a decade and it worked fine. Only within the last year did we (Harvard Dataverse) switch to S3. 😄

**Comment by poikilotherm** [comment]
Thx! Joined :smile:

What approach would you guys prefer? Separate driver with custom URL prefix or an extension of the existing driver?

**Comment by pdurbin** [comment]
@poikilotherm thanks. I'll mention this issue at standup in about an hour but I guess I'm thinking a separate driver would be a good place to start. So rather than S3AccessIO.java it would be called MinioAccessIO.java maybe? I hope I'm understanding what you're building. I think it's a Minio driver.

**Comment by matthew-a-dunlap** [comment]
I am fairly sure supporting custom endpoints would be adding one line to the s3AccessIO.java where the client builder is initialized, to take in a custom url. That and a new config variable.

**Comment by poikilotherm** [comment]
@matthew-a-dunlap sure, will do so. Maybe you can give #5068 a look, too? This might be a dependency for this...

**Comment by poikilotherm** [comment]
Hey @matthew-a-dunlap and @pdurbin,

I think I got this ready for review.

Beware of the monsters in the shadows: 

1. I could only test this with [Minio](http://minio.io), not with AWS, as I do not have an AWS account.
2. As I wrote the unit test in JUnit5, this is blocked by #2940. That's why the branch is based on that feature branch, too.
3. As there is no such things as Arquillian in place right now, there is NO integration test. IMHO this is not really possible right now, as it would need a changed docker container with Minio or something else inside. Arquillian to the rescue... :wink: 
4. As I am great JetBrains IDEA fan, I smuggled two commits into this feature branch that adds some stuff to gitignore and fixed the Maven model. Hope you guys don't mind. If you do, I can split and open another PR.

**Comment by poikilotherm** [comment]
@sparrowjack63 and @mbamouni, could you test my feature branch with your S3 services and report back? Thank you!!!

**Comment by pdurbin** [comment]
@poikilotherm thanks! Pull request #5059 is the one to look at, right? (The description still says "do not merge".) I moved this issue over to code review at https://waffle.io/IQSS/dataverse

**Comment by poikilotherm** [comment]
Sorry, missed the description. Thank for pointing that out, fixed. :smile_cat:

**Comment by poikilotherm** [comment]
Hi @matthew-a-dunlap you guys should already have access to the pull request branch by default, haven't you? Or are you asking me to enable push access for @IQSS for the whole fork?

**Comment by matthew-a-dunlap** [comment]
@poikilotherm I was running into issues getting to the point where i could push to the branch, but it may have been a problem on my end

**Comment by poikilotherm** [comment]
It definitly is enabled...
![grafik](https://user-images.githubusercontent.com/5468333/46033930-af987d80-c0ff-11e8-9a0e-d0f7886efc83.png)

**Comment by poikilotherm** [comment]
Hey @sparrowjack63 and @mbamouni,

I would really appreciate if you could test this feature branch and give us some feedback. If this is no longer relevant to you, maybe you can drop a comment anyway? Thanks!!!

**Comment by poikilotherm** [comment]
Hey @djbrooke, @dlmurphy and @matthew-a-dunlap (and maybe @pdurbin ?),

could you guys please approve my PR #5059 so it might get merged soon? That would be really awesome! If there is anything missing or if anything needs to be changed, just drop me a line.

Thank you! You're awesome building Dataverse! :smile: :+1:

**Comment by pdurbin** [comment]
@poikilotherm our convention is that while an issue is in Code Review at https://waffle.io/IQSS/dataverse the assignees on the issue indicate who is still doing work on it. So if you're done working on the issue, can I unassign you? It will make the status of the issue more clear at standup in about an hour.

@dlmurphy you're still assigned as well from back when you self-assigned last week. I assume you'll take yourself off when you're done making changes.

**Comment by dlmurphy** [comment]
I'm all set, moving this to QA.

**Comment by kcondon** [comment]
The underlying PR is blocked so awaiting clarification on status.

**Comment by djbrooke** [comment]
Thanks @poikilotherm - I'll move it over to QA.

**Comment by poikilotherm** [comment]
Hey @kcondon,
is there any chance this could be merged anytime soon?
Thx!

**Comment by kcondon** [comment]
@poikilotherm  Sorry, I have been a bit backlogged. I'm working on this now.

**Comment by poikilotherm** [comment]
Thanks for fixing the docs @pdurbin and thanks for merging @kcondon!

Glad this is in upstream now :tada:

**Comment by pdurbin** [comment]
@poikilotherm sure. @kcondon and I were talking about improvements we could make to the docs in the future but we didn't want to hold up merging just for docs. Thank you for your contribution!

**Comment by poikilotherm** [comment]
@pdurbin I just saw this issue is not assigned to a milestone... Should this be added to a 4.9.5 or 4.10 release?

**Comment by pdurbin** [comment]
@poikilotherm if I knew for sure we were planning to release 4.9.5 I would add it there but I don't what the next release will be or when. A number of 4.x releases are mentioned at https://dataverse.org/goals-roadmap-and-releases
